### PR TITLE
OPDS name mapper

### LINKED ISSUE #828: OPDS catalog must respect the used nameMapper
**Describe the bug**
On libkiwix 10.1.1 , zim file while opening over hotspot resulting ```The requested URL "/" was not found on this server.```.

**Expected behavior**
Its should open the zim file without any issue

**Screenshots**
Screenshot from 2022-10-03 18-52-57
Screenshot from 2022-10-03 18-53-01

**Environment**
- Version of Kiwix Android : 3.7.0
- Device : REDMI NOTE 9
- OS version : 11

### PR DESCRIPTION
**Pull request by mgautierfr**
Fixes #828

### CODE REVIEW

File: src/opds_dumper.cpp [hunk 1]
```diff
@@ -94,24 +95,24 @@ kainjow::mustache::object getSingleBookData(const Book& book)
     };
 }
 
-std::string getSingleBookEntryXML(const Book& book, bool withXMLHeader, const std::string& rootLocation, const std::string& endpointRoot, bool partial)
+std::string getSingleBookEntryXML(const Book& book, const NameMapper& nameMapper,  bool withXMLHeader, const std::string& rootLocation, const std::string& endpointRoot, bool partial)
```
Comment (veloman-yunkan) [inline src/opds_dumper.cpp:98]: I started #829 in order to avoid ending up with such a function receiving 6 parameters.

Comment (mgautierfr) [reply src/opds_dumper.cpp:98]: Yes, that's why I've started a factorization after this commit :)

File: include/server.h [hunk 2]
```diff
@@ -29,14 +29,85 @@ namespace kiwix
   class NameMapper;
   class InternalServer;
 
+  class ServerConfiguration {
+    public:
+      ServerConfiguration(Library* library, NameMapper* nameMapper=nullptr)
```
Comment (veloman-yunkan) [inline include/server.h]: The second parameter being optional, `explicit` is recommended here

File: include/server.h [hunk 3]
```diff
@@ -29,14 +29,85 @@ namespace kiwix
...
+
   class Server {
      public:
        /**
         * The default constructor.
         *
         * @param library The library to serve.
         */
-       Server(Library* library, NameMapper* nameMapper=nullptr);
+       Server(const ServerConfiguration& configuration);
```
Comment (veloman-yunkan) [inline include/server.h:107]: `explicit` was due even before this change

File: include/server.h [hunk 4]
```diff
@@ -29,14 +29,85 @@ namespace kiwix
   class NameMapper;
   class InternalServer;
 
+  class ServerConfiguration {
```
Comment (veloman-yunkan) [inline include/server.h:32]: What about making this a nested class of `Server` (`Server::Configuration`)?

File: src/server/internalServer.cpp [hunk 5]
```diff
@@ -406,13 +385,13 @@ bool InternalServer::start() {
...
 #endif
-  if (m_verbose.load())
+  if (m_configuration.m_verbose)
     flags |= MHD_USE_DEBUG;
 
   struct sockaddr_in sockAddr;
   memset(&sockAddr, 0, sizeof(sockAddr));
   sockAddr.sin_family = AF_INET;
-  sockAddr.sin_port = htons(m_port);
+  sockAddr.sin_port = htons(m_configuration.m_port);
```
Comment (veloman-yunkan) [inline src/server/internalServer.cpp:380]: You could have avoided the numerous changes like this one by making `InternalServer` a (private) subclass of `ServerConfiguration`.

File: test/library.cpp [hunk 6]
```diff
@@ -236,14 +236,14 @@ namespace
...
+  auto lib = std::make_shared<kiwix::Library>();
+  kiwix::Manager manager(lib);
   manager.readOpds(sampleOpdsStream, "library-opds-import.unittests.dev");
 
-  EXPECT_EQ(10U, lib.getBookCount(true, true));
+  EXPECT_EQ(10U, lib->getBookCount(true, true));
 
   {
-  const kiwix::Book& book1 = lib.getBookById("0c45160e-f917-760a-9159-dfe3c53cdcdd");
+  const kiwix::Book& book1 = lib->getBookById("0c45160e-f917-760a-9159-dfe3c53cdcdd");
```
Comment (veloman-yunkan) [inline test/library.cpp:246]: You could have avoided the numerous changes like this one by utilizing a `shared_ptr` with a non-deleting deallocator:

https://github.com/kiwix/libkiwix/blob/a844bc4000de0cce1e0b5a1b0b5356a3160fd99a/src/manager.cpp#L33-L36

https://github.com/kiwix/libkiwix/blob/a844bc4000de0cce1e0b5a1b0b5356a3160fd99a/src/manager.cpp#L92-L102

File: src/opds_dumper.cpp [hunk 7]
```diff
@@ -99,12 +100,12 @@ std::string fullEntryXML(const Book& book, const std::string& rootLocation, cons
     return render_template(RESOURCE::templates::catalog_v2_entry_xml, data);
 }
 
-std::string partialEntryXML(const Book& book, const std::string& rootLocation)
+std::string partialEntryXML(const ServerConfiguration& configuration, const Book& book)
```
Comment (veloman-yunkan) [inline src/opds_dumper.cpp:103]: I don't agree with this change.

Cons:
- access to more data than needed by this function

Pros:
- ???

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/opds_dumper.cpp [hunk 8]
```diff
@@ -71,7 +72,7 @@ IllustrationInfo getBookIllustrationInfo(const Book& book)
     return illustrations;
 }
 
-std::string fullEntryXML(const Book& book, const std::string& rootLocation)
+std::string fullEntryXML(const Book& book, const std::string& rootLocation, const std::string& bookName)
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/opds_dumper.cpp]: Update: I see that this concern was later on addressed in a slightly different way.

File: src/opds_dumper.cpp (hunk 7 above)
Comment (mgautierfr) [reply src/opds_dumper.cpp:103]: I made this for consistency.
The idea is that those set of free functions to generate xml are black box function generating the xml in accordance with the ServerConfiguration, what ever is actually used.

I have no strong opinion on this,except that passing the actual needed data may leads to too much parameters.
And if several functions are using a ServerConfiguration I prefer having all functions using them.

**Comment by mgautierfr** [comment]
Ready for a new review.
New commits are marked with `[NEW]` (or `fixup!` as usual)

File: src/server/internalServer.cpp [hunk 9]
```diff
@@ -216,7 +216,7 @@ std::pair<std::string, Library::BookIdSet> InternalServer::selectBooks(const Req
   try {
     auto bookName = request.get_argument("content");
     try {
-      const auto bookIds = Library::BookIdSet{m_configuration.mp_nameMapper->getIdForName(bookName)};
+      const auto bookIds = Library::BookIdSet{mp_nameMapper->getIdForName(bookName)};
```
Comment (veloman-yunkan) [inline src/server/internalServer.cpp:219]: This comment applies to the commit "[NEW] Make InternalServer inherite Server::Configuration.". It would be better to follow this approach from the very beginning, instead of making a lot of changes and then undoing them.

**Review by veloman-yunkan — COMMENTED** [review]
Please rebase&fixup. I think that the clean version has to be reviewed due to significant reversal of some changes.

**Review by veloman-yunkan — COMMENTED** [review]
While reviewing kiwix/kiwix-tools#580 and kiwix/kiwix-desktop#894 I got the impression that this PR goes too far with passing `Library` and `NameMapper` via shared pointers into some objects.

`OPDSDumper` and `SearchRenderer` are cheap-to-instantiate one-shot-use helper classes with the (current) mere purpose of avoiding free functions `dumpOPDSFeed()`, `renderSearchResults()` with too many parameters. Objects of those classes should be created, used and destroyed within a single function call, and therefore storing the references to `Library` and `NameMapper` as shared pointers is not justified. In case of `OPDSDumper` this concern is somewhat masked by replacing individual references to `Library` and `NameMapper` by an entire `Server::Configuration` object, but the result is that `OPDSDumper` gains non-const access to `Library` which it doesn't really need.

Regarding `Manager`, a scenario where it may turn out to be the last object keeping a reference to `Library` doesn't make much sense (at least, for the case of a manager object instantiated via the `Manager(Library*)` constructor). In such a situation, the modification of the library via the `Manager` should have ended before other clients of the library die.

**Comment by mgautierfr** [comment]
I was more concern by the `Server::Configuration` (and so the server) keeping a reference to a destroyed `Library`.

But I agree that `OPDSDumper` and `SearchRendered` become a bit too complex. If we agree that those objects are one-shoot helper, I would tend to move them as free functions (taking a `Configuration` object if there is too many parameters) and keep the class "private". This way it would not be possible to keep an existing instance having a reference to a `Library`.

There is also the `(Updatable)NameMapper` keeping a reference to the `Library`. We may fix that by having the `Library` keeping a reference to the name mapper instead. It would extend `Library` for something that is related only to the server but is should be ok.

This may me think to https://github.com/kiwix/libkiwix/issues/740. `OPDSDumper` and `SearchRendered` are also used in a context of a "Server" (has they need to know how to generate urls)

---

I'm afraid this PR will become more complex to solve that I was initially thinking. I will move the actual fix in another PR to have it quickly merged and will take time to discuss the best design here.

**Comment by veloman-yunkan** [comment]
@mgautierfr

While reviewing #837, I had a new idea. Why do we need to pass `NameMapper` so deep into various operations having to deal with the book content-id? Why don't we replace `Book::getHumanReadableIdFromPath()` with a new user-settable field `Book::m_human_friendly_id` that by default will be initialized to a value previously returned by `getHumanReadableIdFromPath()`? Then name-mapping can be applied to the library in a single place and all clients of the human-friendly id will access it via the enhanced interface of `Book` without needing an extra mapper object. Of course this gives up some flexibility (for example, you can't have different mappings in different contexts for the same library) but do we really need that?

In short, the proposal is to get rid of the `kiwix::NameMapper` interface altogether, incorporating that info in `kiwix::Book`.

**Comment by mgautierfr** [comment]
> Of course this gives up some flexibility (for example, you can't have different mappings in different contexts for the same library) but do we really need that?

Yes, at least for now.

The namemapper is used for two different (but related) things:
- Routing : To know which zim file to use for a request. And several "request" can lead to same zim file (as date alias).
- Generating content : To create links which must of course be in sync with the routing.

The problem is that routing is not part of the "whole" library (libkiwix). It is part of server only. kiwix-desktop has it own routing and it use the namemapper to generate search result in correspondance to its routing. In the same time, kiwix-desktop can serve the same library (set of books) with kiwix-serve (and so with a different namemapper). And it is the same thing for kiwix-android (but I'm not sure kiwix-android is generating search result with `searchRenderer`).

When we implement #740 with some kind of generic routing system, we may rethink this as we could provide of full set (consistant) of features.
